src/theory/ff/theory_ff_rewriter.cpp

```diff
@@ -15,20 +15,217 @@
 {
   Trace("ff::rw::post") << "ff::postRewrite: " << t << std::endl;
-  return RewriteResponse(REWRITE_DONE, t);
+  switch (t.getKind())
+  {
+    case kind::FINITE_FIELD_NEG: return RewriteResponse(REWRITE_DONE, t);
+    case kind::FINITE_FIELD_ADD:
+      return RewriteResponse(REWRITE_DONE, postRewriteFfAdd(t));
+    case kind::FINITE_FIELD_MULT:
+      return RewriteResponse(REWRITE_DONE, postRewriteFfMult(t));
+    case kind::EQUAL: return RewriteResponse(REWRITE_DONE, postRewriteFfEq(t));
+    default: return RewriteResponse(REWRITE_DONE, t);
+  }
 }
 
 RewriteResponse TheoryFiniteFieldsRewriter::preRewrite(TNode t)
 {
   Trace("ff::rw::pre") << "ff::preRewrite: " << t << std::endl;
-  return RewriteResponse(REWRITE_DONE, t);
+  switch (t.getKind())
+  {
+    case kind::FINITE_FIELD_NEG:
+      return RewriteResponse(REWRITE_DONE, preRewriteFfNeg(t));
+    case kind::FINITE_FIELD_ADD:
+      return RewriteResponse(REWRITE_DONE, preRewriteFfAdd(t));
+    case kind::FINITE_FIELD_MULT:
+      return RewriteResponse(REWRITE_DONE, preRewriteFfMult(t));
+    case kind::EQUAL: return RewriteResponse(REWRITE_DONE, t);
+    default: return RewriteResponse(REWRITE_DONE, t);
+  }
+}
+
+/** preRewrite negation */
+Node preRewriteFfNeg(TNode t)
+{
+  Assert(t.getKind() == Kind::FINITE_FIELD_NEG);
+  NodeManager* const nm = NodeManager::currentNM();
+  const Node negOne = nm->mkConst(FfVal(Integer(-1), t.getType().getFfSize()));
+  return nm->mkNode(kind::FINITE_FIELD_MULT, negOne, t[0]);
+}
+
+/** preRewrite addition */
+Node preRewriteFfAdd(TNode t)
+{
+  Assert(t.getKind() == Kind::FINITE_FIELD_ADD);
+  return expr::algorithm::flatten(t);
+}
+
+/** postRewrite addition */
+Node postRewriteFfAdd(TNode t)
+{
+  const TypeNode field = t.getType();
+  Assert(field.isFiniteField());
+
+  FfVal one = FfVal::mkOne(field.getFfSize());
+
+  FfVal constantTerm = FfVal::mkZero(field.getFfSize());
+  std::map<Node, FfVal> scalarTerms;
+
+  std::vector<TNode> children;
+  expr::algorithm::flatten(t, children);
+
+  for (const auto& child : children)
+  {
+    if (child.isConst())
+    {
+      constantTerm = constantTerm + child.getConst<FfVal>();
+    }
+    else
+    {
+      std::optional<std::pair<Node, FfVal>> parsed = parseScalar(child);
+      std::pair<Node, FfVal> pair = parsed.value_or(std::make_pair(child, one));
+      const auto entry = scalarTerms.find(pair.first);
+      if (entry == scalarTerms.end())
+      {
+        scalarTerms.insert(pair);
+      }
+      else
+      {
+        entry->second = entry->second + pair.second;
+      }
+    }
+  }
+  NodeManager* const nm = NodeManager::currentNM();
+  std::vector<Node> summands;
+  if (scalarTerms.empty() || !constantTerm.getValue().isZero())
+  {
+    summands.push_back(nm->mkConst(constantTerm));
+  }
+  for (const auto& summand : scalarTerms)
+  {
+    if (summand.second.getValue().isZero())
+    {
+      // drop this term
+    }
```

Comment:
Can this happen? The children should already be rewritten at this point, so there shouldn't be something like `0 * x` as a child.

Reply:
This is a bit subtle. It can happen b/c of cancellation, consider `(+ (+ x y) (+ (* -1 x) z))`.

I've added a comment to this effect.